Svc/FpySequencer/FpySequencerStateMachine.cpp

```diff
@@ -0,0 +1,313 @@
+) {
+    bool result = this->validate();
+    if (!result) {
+        this->sequencer_sendSignal_result_failure();
+        return;
+    }
+    this->sequencer_sendSignal_result_success();
+}
+
+//! Implementation for action checkShouldWake of state machine Svc_FpySequencer_SequencerStateMachine
+//!
+//! checks if sequencer should wake from sleep
+void FpySequencer::Svc_FpySequencer_SequencerStateMachine_action_checkShouldWake(
+    SmId smId,                                             //!< The state machine id
+    Svc_FpySequencer_SequencerStateMachine::Signal signal  //!< The signal
+) {
+    Fw::Time currentTime = this->getTime();
+
+    if (currentTime.getTimeBase() != m_runtime.wakeupTime.getTimeBase()) {
+        // cannot compare these times.
+        this->log_WARNING_LO_MismatchedTimeBase(currentTime.getTimeBase(), m_runtime.wakeupTime.getTimeBase());
+
+        this->sequencer_sendSignal_result_timeOpFailed();
+        return;
+    }
+
+    if (currentTime.getContext() != m_runtime.wakeupTime.getContext()) {
+        // cannot compare these times.
+        this->log_WARNING_LO_MismatchedTimeContext(currentTime.getContext(), m_runtime.wakeupTime.getContext());
+
+        this->sequencer_sendSignal_result_timeOpFailed();
+        return;
+    }
+
+    if (currentTime < m_runtime.wakeupTime) {
+        // not time to wake up!
+        return;
+    }
+
+    // say we've finished our sleep
+    this->sequencer_sendSignal_result_checkShouldWake_wakeup();
+}
+
+//! Implementation for action checkStatementTimeout of state machine Svc_FpySequencer_SequencerStateMachine
+//!
+//! checks if the current statement has timed out
+void FpySequencer::Svc_FpySequencer_SequencerStateMachine_action_checkStatementTimeout(
+    SmId smId,                                             //!< The state machine id
+    Svc_FpySequencer_SequencerStateMachine::Signal signal  //!< The signal
+) {
+    Fw::ParamValid valid;
+    F32 timeout = this->paramGet_STATEMENT_TIMEOUT_SECS(valid);
+    if (timeout <= 0) {
+        // no timeout
+        return;
+    }
+
+    Fw::Time currentTime = getTime();
+
+    if (currentTime.getTimeBase() != m_runtime.currentStatementDispatchTime.getTimeBase()) {
+        // can't compare time base. must have changed
+        this->log_WARNING_LO_MismatchedTimeBase(currentTime.getTimeBase(),
+                                                m_runtime.currentStatementDispatchTime.getTimeBase());
+        this->sequencer_sendSignal_result_timeOpFailed();
+        return;
+    }
+    if (currentTime.getContext() != m_runtime.currentStatementDispatchTime.getContext()) {
+        // can't compare time ctx. must have changed
+        this->log_WARNING_LO_MismatchedTimeContext(currentTime.getContext(),
+                                                   m_runtime.currentStatementDispatchTime.getContext());
+        this->sequencer_sendSignal_result_timeOpFailed();
+        return;
+    }
+
+    F64 currentTimeSecs =
```

Comment:
Seems odd to need a F64 for these calculations. Could probably subtract the seconds and sub-seconds independently, then convert to the F32.

This would be nicer on systems with soft double support.